src/qililab/qprogram/qblox_compiler.py

```diff
@@ -464,14 +473,35 @@ def _handle_add_waits(self, bus: str, duration: int):
             duration (int): duration to wait in ns
         """
         if duration > INST_MAX_WAIT:
+            if self._buses[bus].upd_param_instruction_pending:
+                self._buses[bus].qpy_block_stack[-1].append_component(component=QPyInstructions.UpdParam(4))
+                duration = duration - 4  # Remove the 4 ns that have been added with the UpdParam
+                self._buses[bus].upd_param_instruction_pending = False
+
             for _ in range(duration // INST_MAX_WAIT):
                 self._buses[bus].qpy_block_stack[-1].append_component(
                     component=QPyInstructions.Wait(wait_time=INST_MAX_WAIT)
                 )
-        # add the remaining wait time (or all of it if the above conditional is false)
-        self._buses[bus].qpy_block_stack[-1].append_component(
-            component=QPyInstructions.Wait(wait_time=duration % INST_MAX_WAIT)
-        )
+            # add the remaining wait time
+            self._buses[bus].qpy_block_stack[-1].append_component(
+                component=QPyInstructions.Wait(wait_time=duration % INST_MAX_WAIT)
+            )
+        else:
+            if self._buses[bus].upd_param_instruction_pending:
+                if duration >= 8:
+                    self._buses[bus].qpy_block_stack[-1].append_component(component=QPyInstructions.UpdParam(4))
+                    self._buses[bus].qpy_block_stack[-1].append_component(
+                        component=QPyInstructions.Wait(wait_time=(duration - 4))
+                    )
+                elif duration == 4:  # replace the wait with an update parameter of 4 ns
+                    self._buses[bus].qpy_block_stack[-1].append_component(component=QPyInstructions.UpdParam(4))
+
+                elif 4 < duration < 8:  # increase the min wait time of the update parameter and remove the wait
+                    self._buses[bus].qpy_block_stack[-1].append_component(component=QPyInstructions.UpdParam(duration))
+
+                self._buses[bus].upd_param_instruction_pending = False
+            else:
+                self._buses[bus].qpy_block_stack[-1].append_component(component=QPyInstructions.Wait(wait_time=duration % INST_MAX_WAIT))
```

Comment:
Is there a way to simplify this a bit? Seems a little confusing with the nested-ifs.